### ISSUE TITLE
Smartcasting sharp edges - tracking issue

### ISSUE THREAD

**Issue by dumblob** [vlang/v#9305]
1. all the points from https://github.com/vlang/ui/pull/331#discussion_r593932159 and https://github.com/vlang/ui/pull/331#discussion_r593899795 (double/triple/... `mut` required; disallow `mut` smartcasting over an immutable variable)
2. a hack around not being able to pass an `ast.CharLiteral | ast.FloatLiteral | ast.IntegerLiteral` to `expr_string`, which expects an `ast.Expr` - see https://github.com/vlang/v/pull/10058#discussion_r629541486
3. unable to assign to a sum type common field: https://github.com/vlang/v/issues/10026
4. flow-sensitive sum type resolution sometimes doesn't work: https://github.com/vlang/v/issues/10379
5. smartcasting has problems with automated dereferencing: https://github.com/vlang/v/issues/11543

**Comment by ntrel** [comment]
If a variable/field is `mut` then smartcasting should not be done because it is memory-unsafe e.g. for pointers. Otherwise the original can be reassigned whilst we still have a shallow copy in scope. Kotlin disallows mutable smart casting for this reason.

**Comment by dumblob** [comment]
As @danieldaeschle pointed out, everything probably boils down to cases where smartcasting seems (or **is really?**) undecidable and thus the smartcasting "branch" has to be chosen explicitly. Maybe a new keyword substituting `is` would be the way to go? Maybe `isin` or `istoo` or `isdeep`...

**Comment by danieldaeschle** [comment]
`is!`?

**Comment by dumblob** [comment]
`is!` was my first idea, but I wasn't sure whether its negation `!is!` looks still readable and understandable. Maybe `isis` (two `is` are something more than one `is` :laughing:).

**Comment by danieldaeschle** [comment]
@dumblob for better understanding it should be here: https://github.com/vlang/v/blob/master/doc/docs.md#smart-casting

**Comment by dumblob** [comment]
Well, the doc just says how I originally understood smartcasting. If that semantic (and no other) is expected, then I'd advocate for the following:
```v
w := ...
if w is Mars {
	assert typeof(w).name == 'Mars'
	if w.dust_storm() {
		println('bad weather!')
	}
}
```

```v
mut w := ...
if w is Mars {
	assert typeof( unsafe{w} ).name == 'Mars'
	if unsafe{w}.dust_storm() {
		println('bad weather!')
	}
}
```
No need for new keywords, no new syntax, nothing. If `unsafe{}` is missing, compiler warns about it. Do you find this also easy to use and crystal clear?

**Comment by EPonyA** [comment]
> 	if unsafe{w}.dust_storm() {

Are curly braces necessary here? It seems to me that V has gradually been tracing Rust (post 1.0) in that it becomes increasingly liberal with its unsafe semantics. Ever more simple features are being added/considered that make unsafe one-liners seem fairly common. How often do you see a major contributor say something like, "I guess we could allow it in unsafe"? Off the top of my head, bool-to-int casting and initialization of built-in types have also been considered for legal ``unsafe`` code, and neither of those operations span multiple lines.

``const`` no longer requires curly braces, which I suppose is because ``const`` one-liners are common and too many braces pollutes code. The justification, afaik, has been that ``unsafe`` code is discouraged, and simple things shouldn't need it, but that seems less reasonable imho while V's ``unsafe`` capabilities continue growing more and more useful.

Or maybe a feature for casting these types with some error handling could be considered instead?

**Comment by dumblob** [comment]
@EPonyA I think it's perfectly OK as it exactly uncovers leaky abstractions higher-level concepts have (and it's good they're leaky otherwise we'd be building another Haskell!).

It's of course possible to write non-`unsafe{}` code, but it'll really look like Lisp code. It's of course also possible to write all-`unsafe{}` code, but it'll really look more like ASM :wink:.

Great languages allow both. Both equally well.

**Comment by danieldaeschle** [comment]
```v
x := unsafe { y } // casted one
x2 = y // uncasted on
```

Will that work then?

**Comment by danieldaeschle** [comment]
But then we can cast it manually anyway:
```v
x := y as MyType
```

**Comment by dumblob** [comment]
> ```v
> x := unsafe { y } // casted one
> x2 = y // uncasted on
> ```
> 
> Will that work then?

Assuming you meant either of these:

```v
y := ...
if y is MyType {
  x := unsafe { y } // casted one
  x2 := y // uncasted on
}
```
^^^ will not work (if we want to be really strict). The first line will produce error, because `unsafe{}` is not needed (but I'm fine if this restriction is alleviated). The second line will compile without error.

```v
mut y := ...
if y is MyType {
  x := unsafe { y } // casted one
  x2 := y // uncasted on
}
```
^^^ will not work. The first line will compile without any error, because `unsafe{}` is required. The second line will not compile and produce error (because `unsafe{}` is missing).

**Comment by danieldaeschle** [comment]
as i said, `x := unsafe { y }` is equivalent to the current `x := y as MyType`.

The second `x2 := y` should still work and give no error. But it's not casted.

**Comment by dumblob** [comment]
> as i said, `x := unsafe { y }` is equivalent to the current `x := y as MyType`.

Hm, I'd actually expect `x := unsafe { y }` to be equivalent to `x := unsafe{ y as MyType }`.

**Comment by danieldaeschle** [comment]
`x as MyType` works without `unsafe`

**Comment by dumblob** [comment]
> `x as MyType` works without `unsafe`

Why is it any different from `if y is MyType { ...`?

This everything sounds flaky and unfinished... Are there any plans to revise these discrepancies?

**Comment by danieldaeschle** [comment]
I hope it will be revisited. I'm not 100% statisfied too.

`as` is a cast `is` is a check with a smartcast if the type is not mutable.

**Comment by dumblob** [comment]
@danieldaeschle actually I was asking about the difference with regards to safety - i.e. why is `unsafe{}` required in one case and not in the other (you said above, that `x := unsafe { y }` is equivalent to the current `x := y as MyType`).

Why this difference in (un)safety? I still can't pinpoint the culprit :cry:.

**Comment by danieldaeschle** [comment]
Yeah, tbh. it should be wrapped in `unsafe`. But it wasn't introduced yet.

**Comment by dumblob** [comment]
Oh, thanks! That explains it. Once you have some more information about decisions etc. regarding smartcasting, please keep me posted (it's enough to write it here down - I'm subscribed to this issue). Thanks a lot!

**Comment by medvednikov** [comment]
V's had smartcasting for a while.

**Comment by danieldaeschle** [comment]
Will replace `if mut x is MyType {´ with `if unsafe { x is Mytype }`. If you omit unsafe it will do the check but won't cast. It's very implicit. But I don't think any compiler hints are possible currently.
